OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -471,8 +471,10 @@ public void SetPosition(Actor self, CPos cell, SubCell subCell = SubCell.Any)
 			var position = cell.Layer == 0 ? self.World.Map.CenterOfCell(cell) :
 				self.World.GetCustomMovementLayers()[cell.Layer].CenterOfCell(cell);
 
-			var subcellOffset = self.World.Map.Grid.OffsetOfSubCell(subCell);
-			SetCenterPosition(self, position + subcellOffset);
+			position += self.World.Map.Grid.OffsetOfSubCell(subCell);
+			position -= new WVec(0, 0, self.World.Map.DistanceAboveTerrain(position).Length);
```

Comment:
I notice the other overload below has some differences. It *adds* the distance above terrain, but you *subtract* here. The other overload also doesn't account for layers. I feel like that other overload has some bugs!

Maybe have the other overload call into this one, so they both benefit from the fix:
```c#
public void SetPosition(Actor self, WPos pos)
{
	var cell = self.World.Map.CellContaining(pos);
	SetPosition(self, cell);
}
```

Reply:
The other overload gets the correct Z value from CenterOfSubCell (rather than using CenterOfCell and then shifting by the subcell offsets). The addition of the distance over ground is done to preserve any existing value that might have existed before setting the position, while the overload I modified would ignore any existing distance over ground and set the Z value to the center of cell height.

It's a bit weird how different the two overloads are, I only changed the above one as the other one gets the height right and I wanted to keep the possible regressions as few as possible.